test/testsuites.yaml

```diff
@@ -0,0 +1,75 @@
+# Select test suites to be run. This list is used together with Git diff.
+# run_if_changed defines path expressions (regexp) and associated test suites to be run
+# if those paths are changed.
+
+# A testsuite without "tests" element defines path expressions that don't require
+# any tests to be run on change.
+
+# Each path in the Git diff is checked against all "run_if_changed" regular expressions.
+# If the path matches any of the expressions then all tests from that testsuite are added
+# to the list of tests to be run.
+# If the path doesn't match any expression in any testsuite then it's an unknown
+# change with a potentially high impact and all tests should run.
+testsuites:
+  - name: "Serverless Operator E2E"
+    run_if_changed:
+      - ^knative-operator/pkg/controller/knativeeventing/
+      - ^knative-operator/pkg/controller/knativeserving/
+      - ^knative-operator/pkg/webhook/knativeeventing/
+      - ^knative-operator/pkg/webhook/knativeserving/
+      - ^openshift-knative-operator/cmd/operator/kodata/knative-eventing/
+      - ^openshift-knative-operator/cmd/operator/kodata/knative-serving/
+      - ^openshift-knative-operator/pkg/eventing/
+      - ^openshift-knative-operator/pkg/serving/
+      - ^serving/ingress/
+    tests:
+      - serverless_operator_e2e_tests
+  - name: "Eventing Kafka"
+    run_if_changed:
+      - "^knative-operator/pkg/controller/knativekafka/"
+      - "^knative-operator/pkg/webhook/knativekafka/"
+    tests:
+      - serverless_operator_kafka_e2e_tests
+      - downstream_knative_kafka_e2e_tests
+      - upstream_knative_eventing_kafka_e2e # TODO: Run this at all? Should this test suite be removed completely?
+      - upstream_knative_eventing_kafka_broker_e2e
+  - name: "Eventing"
+    run_if_changed:
+      - "^knative-operator/pkg/controller/knativeeventing/"
+      - "^knative-operator/pkg/webhook/knativeeventing/"
+      - "^openshift-knative-operator/cmd/operator/kodata/knative-eventing/"
+      - "^openshift-knative-operator/pkg/eventing/"
+    tests:
+      - downstream_eventing_e2e_tests
+      - upstream_knative_eventing_e2e
+  - name: "Monitoring"
+    run_if_changed:
+      - "^knative-operator/pkg/monitoring/"
+      - "^openshift-knative-operator/cmd/operator/kodata/monitoring/"
+      - "^openshift-knative-operator/pkg/monitoring/"
+    tests:
+      - downstream_monitoring_e2e_tests
+  - name: "Serving"
+    run_if_changed:
+      - "^knative-operator/pkg/controller/knativeserving/"
+      - "^knative-operator/pkg/webhook/knativeserving/"
+      - "^openshift-knative-operator/cmd/operator/kodata/knative-serving/"
+      - "^openshift-knative-operator/pkg/serving/"
+      - "^serving/ingress/"
+    tests:
+      - downstream_serving_e2e_tests
+  - name: "UI"
+    run_if_changed:
+      - "^ui/"
```

Comment:
```suggestion
      - "^test/ui/"
```
right?

Reply:
Yes. Anyway, we run the UI tests in separate Prow check and I didn't include the "should run/should not run" logic in the UI tests script yet.
I might as well remove this UI section from this testsuites.yaml file. Let's discuss later.

Reply:
@cardil Let's decide what we want to do with UI tests.
In theory, we could get rid of this [section of ci-operator config](https://github.com/openshift/release/blob/master/ci-operator/config/openshift-knative/serverless-operator/openshift-knative-serverless-operator-main__4.12.yaml#L378-L405).

We could run UI tests as part of [operator-e2e](https://github.com/openshift/release/blob/master/ci-operator/config/openshift-knative/serverless-operator/openshift-knative-serverless-operator-main__4.12.yaml#L117).

The behavior would be following:
If the changed paths match only `^test/ui/` the UI tests would run.
If the changed paths match anything else from testsuites.yaml as well as `^test/ui/` the UI tests would run as well (along other selected tests).
If the changed paths match `^test/ui/` and then something that is not defined in testsuites.yaml then UI tests would NOT run. In this case it would run "all" tests except for UI tests.

Examples:
Changed paths:
```
test/ui/package.json
``` 
Result: Run UI tests.

Changed paths:
```
test/ui/package.json
hack/lib/serverless.bash
```
Result: Do NOT run UI tests.

This behaviour is a little different that what we have now with UI tests. Now they run if any changed path matches `test/ui`, other paths can match whatever else but the UI tests will still run.

Are you ok with this?

The other option would be to run UI tests even in this latest case. But that would mean UI tests would run with most pull requests because usually there are at least some changes which do not match paths from testsuites.yaml.

Reply:
Anyway, I'm not convinced about this. Maybe the UI tests should be treated separately due to their instability. We could leave them as they are now.

Reply:
Maybe a good option would be to merge ci-operator configs for operator-e2e and serverless-ui into one:
```
- as: operator-e2e-aws-ocp-412
  cluster_claim:
    architecture: amd64
    cloud: aws
    owner: openshift-ci
    product: ocp
    timeout: 2h0m0s
    version: "4.12"
  steps:
    allow_skip_on_success: true
    test:
    - as: operator-e2e
      commands: make test-e2e-with-kafka
      dependencies:
      - env: KNATIVE_OPENSHIFT_INGRESS
        name: knative-openshift-ingress
      - env: KNATIVE_OPERATOR
        name: knative-operator
      - env: OPENSHIFT_KNATIVE_OPERATOR
        name: openshift-knative-operator
      from: serverless-operator-src
      resources:
        limits:
          memory: 6Gi
        requests:
          cpu: 100m
          memory: 200Mi
    - as: ui
      commands: make test-ui-e2e-testonly
      from: serverless-operator-src
      resources:
        limits:
          memory: 6Gi
        requests:
          cpu: 100m
          memory: 200Mi
    workflow: generic-claim
```

Then we could use our own filtering and have UI tests executed when anything changes under `^test/ui`. The other changed paths would not matter. 
This way we'd get the same behaviour as we have now. Just with a shorter CI operator config.

Reply:
I've pushed a commit that supports this.

Reply:
Hmm, merging the ci-operator config like in the example would still make the UI tests run too often. We don't want that. Maybe better to keep them separate.

Reply:
The changes I have committed so far allow for selecting UI tests based on changes in a PR. But they don't have effect now because we have the separate ci-operator config and we run them separately (not as part of the main test suite). At this point I think it's best to not include UI tests in the automatic test selection and revert this part.